File: dlp_web_only/secure_config.py
"""
===============================================================================
SECURE CONFIG LOADER
===============================================================================
Safely load and manage environment variables and secrets for the application.
This module ensures that secrets are never hardcoded and provides a secure,
centralized way to access configuration values.

Usage:
    from secure_config import config, Config
    api_key = config.get_secret('DLP_API_KEY')
    port = config.get_int('SERVER_PORT', default=5000)

===============================================================================
SECURITY BEST PRACTICES IMPLEMENTED:
===============================================================================
1. ✓ Loads secrets from .env files (never hardcoded)
2. ✓ Validates required secrets are present
3. ✓ Never logs or prints secret values
4. ✓ Provides typed access to configuration values
5. ✓ Supports multiple environment files (.env, .env.local, etc.)
6. ✓ Validates configuration on startup
7. ✓ Type hints for better IDE support
8. ✓ Clear error messages when secrets are missing

===============================================================================
"""

import os
import logging
from pathlib import Path
from typing import Optional, List, Any, Union
from functools import lru_cache

try:
    from dotenv import load_dotenv
except ImportError:
    raise ImportError(
        "python-dotenv is required. Install it with: pip install python-dotenv"
    )

# Configure logging
logger = logging.getLogger(__name__)


class ConfigurationError(Exception):
    """Raised when configuration is missing or invalid"""
    pass


class SecureConfig:
    """
    Secure configuration manager for loading environment variables.
    
    Features:
    - Loads from .env files with proper precedence
    - Validates required secrets are present
    - Provides type-safe access to configuration
    - Never logs secret values
    - Supports environment-specific configurations
    """
    
    # Required secrets that must be present in production
    REQUIRED_SECRETS_PRODUCTION = [
        'DLP_API_KEY',
        'FLASK_SECRET_KEY',
        'ALERT_EMAIL',
    ]
    
    # Required secrets for all environments
    REQUIRED_SECRETS_ALL = [
        'DLP_API_KEY',
    ]
    
    def __init__(self, env_files: Optional[List[str]] = None):
        """
        Initialize configuration loader.
        
        Args:
            env_files: List of .env file paths to load. If None, uses defaults.
                      Files are loaded in order; later files override earlier ones.
        """
        if env_files is None:
            env_files = self._get_default_env_files()
        
        self.env_files = env_files
        self.loaded = False
        self._load_env_files()
        self._validate_configuration()
    
    def _get_default_env_files(self) -> List[str]:
        """
        Get default .env file locations in order of precedence.
        Later files override earlier ones.
        """
        env = os.getenv('FLASK_ENV', 'development')
        workspace_root = Path(__file__).parent
        
        files = [
            workspace_root / '.env',
            workspace_root / f'.env.{env}',
            workspace_root / '.env.local',
        ]
        
        # Filter to only existing files
        existing_files = [str(f) for f in files if f.exists()]
        return existing_files
    
    def _load_env_files(self) -> None:
        """Load environment variables from .env files."""
        if not self.env_files:
            logger.warning(
                "⚠️  No .env files found. Create .env file with required secrets. "
                "See .env.example for template."
            )
            return
        
        for env_file in self.env_files:
            try:
                load_dotenv(env_file, override=True)
                logger.info(f"✓ Loaded configuration from: {env_file}")
            except Exception as e:
                raise ConfigurationError(
                    f"Failed to load {env_file}: {str(e)}"
                )
        
        self.loaded = True
    
    def _validate_configuration(self) -> None:
        """Validate that all required secrets are present."""
        env = os.getenv('FLASK_ENV', 'development')
        
        # Always check basic required secrets
        required = self.REQUIRED_SECRETS_ALL
        
        # Check production-specific requirements
        if env == 'production':
            required = self.REQUIRED_SECRETS_PRODUCTION
        
        missing = []
        for secret in required:
            if not os.getenv(secret):
                missing.append(secret)
        
        if missing:
            error_msg = (
                f"Missing required environment variables in {env} mode: "
                f"{', '.join(missing)}\n"
                f"Instructions:\n"
                f"  1. Copy .env.example to .env\n"
                f"  2. Fill in actual values for: {', '.join(missing)}\n"
                f"  3. Never commit .env to version control"
            )
            raise ConfigurationError(error_msg)
    
    def get(self, key: str, default: Optional[str] = None) -> Optional[str]:
        """
        Get a configuration value as string.
        
        Args:
            key: Environment variable name
            default: Default value if not found
        
        Returns:
            Configuration value or default
        
        Note: This method is safe to use even with sensitive values
              as it doesn't log the actual values
        """
        value = os.getenv(key, default)
        if value is None and key in self.REQUIRED_SECRETS_ALL:
            raise ConfigurationError(
                f"Required secret '{key}' not found in environment. "
                f"Set it in .env file."
            )
        return value
    
    def get_secret(self, key: str, required: bool = False) -> str:
        """
        Get a secret value securely.
        
        Args:
            key: Environment variable name for the secret
            required: If True, raises error if secret is not found
        
        Returns:
            Secret value
        
        Raises:
            ConfigurationError: If required=True and secret is missing
        
        Note: Never logs the actual secret value
        """
        value = os.getenv(key)
        if value is None:
            if required:
                raise ConfigurationError(
                    f"Required secret '{key}' not found in environment. "
                    f"Please set it in your .env file."
                )
            return None
        return value
    
    def get_int(self, key: str, default: Optional[int] = None) -> int:
        """
        Get a configuration value as integer.
        
        Args:
            key: Environment variable name
            default: Default value if not found
        
        Returns:
            Configuration value as integer
        
        Raises:
            ConfigurationError: If value cannot be converted to integer
        """
        value = os.getenv(key)
        if value is None:
            if default is not None:
                return default
            raise ConfigurationError(f"Configuration '{key}' not found")
        
        try:
            return int(value)
        except ValueError:
            raise ConfigurationError(
                f"Configuration '{key}' must be an integer, got: {value}"
            )
    
    def get_bool(self, key: str, default: bool = False) -> bool:
        """
        Get a configuration value as boolean.
        
        Args:
            key: Environment variable name
            default: Default value if not found
        
        Returns:
            Configuration value as boolean
        
        Note: Truthy values are: 'true', 'True', '1', 'yes', 'on'
              Falsy values are: 'false', 'False', '0', 'no', 'off', ''
        """
        value = os.getenv(key, str(default)).lower()
        return value in ('true', '1', 'yes', 'on')
    
    def get_list(self, key: str, delimiter: str = ',', default: Optional[List[str]] = None) -> List[str]:
        """
        Get a configuration value as list.
        
        Args:
            key: Environment variable name
            delimiter: Delimiter to split the value (default: comma)
            default: Default list if not found
        
        Returns:
            Configuration value split as list
        """
        value = os.getenv(key)
        if value is None:
            return default or []
        
        return [item.strip() for item in value.split(delimiter) if item.strip()]
    
    def validate_required(self, keys: List[str]) -> None:
        """
        Validate that required keys are present.
        
        Args:
            keys: List of required environment variable names
        
        Raises:
            ConfigurationError: If any required key is missing
        """
        missing = [key for key in keys if not os.getenv(key)]
        if missing:
            raise ConfigurationError(
                f"Missing required configuration: {', '.join(missing)}"
            )
    
    def __repr__(self) -> str:
        """String representation without exposing secrets"""
        env = os.getenv('FLASK_ENV', 'development')
        return (
            f"<SecureConfig "
            f"environment={env} "
            f"loaded_files={len(self.env_files)} "
            f"validation_passed=True>"
        )


# ============================================================================
# SINGLETON INSTANCE
# ============================================================================
# Create a singleton instance for use throughout the application
try:
    config = SecureConfig()
    logger.info("✓ Configuration loaded successfully")
except ConfigurationError as e:
    logger.error(f"✗ Configuration Error: {e}")
    raise


# ============================================================================
# HELPER FUNCTIONS FOR FLASK APPLICATION
# ============================================================================

def init_secure_config(app) -> None:
    """
    Initialize Flask app with secure configuration.
    
    Usage in Flask app:
        from secure_config import init_secure_config
        app = Flask(__name__)
        init_secure_config(app)
    """
    try:
        # Set Flask configuration from environment
        app.config['SECRET_KEY'] = config.get_secret('FLASK_SECRET_KEY', required=True)
        app.config['DEBUG'] = config.get_bool('FLASK_DEBUG', default=False)
        app.config['JSON_SORT_KEYS'] = False
        
        # Security settings
        app.config['PERMANENT_SESSION_LIFETIME'] = config.get_int(
            'SESSION_TIMEOUT', 
            default=1800
        )
        app.config['SESSION_COOKIE_SECURE'] = config.get_bool('USE_HTTPS', False)
        app.config['SESSION_COOKIE_HTTPONLY'] = True
        app.config['SESSION_COOKIE_SAMESITE'] = 'Lax'
        
        logger.info("✓ Flask application configured securely")
    except ConfigurationError as e:
        logger.error(f"Failed to initialize Flask config: {e}")
        raise


# ============================================================================
# EXAMPLE USAGE
# ============================================================================
if __name__ == '__main__':
    """
    Example usage of the SecureConfig module.
    """
    try:
        print("Loading configuration...")
        
        # Access various configuration types
        api_key = config.get_secret('DLP_API_KEY', required=True)
        print("✓ API Key loaded (value hidden for security)")
        
        port = config.get_int('SERVER_PORT', default=5000)
        print(f"✓ Server Port: {port}")
        
        debug = config.get_bool('FLASK_DEBUG', False)
        print(f"✓ Debug Mode: {debug}")
        
        paths = config.get_list('MONITORED_PATHS')
        print(f"✓ Monitored Paths: {paths}")
        
        print("\n✓ All configurations loaded successfully!")
        
    except ConfigurationError as e:
        print(f"✗ Configuration Error: {e}")
        exit(1)

File: dlp_web_only/secure_config.js
"""
===============================================================================
SECURE CONFIG LOADER - NODE.JS VERSION
===============================================================================
Safely load and manage environment variables and secrets for Node.js applications.
This module ensures that secrets are never hardcoded and provides a secure,
centralized way to access configuration values.

Installation:
    npm install dotenv

Usage:
    const config = require('./secure_config.js');
    const apiKey = config.getSecret('API_KEY');
    const port = config.getInt('SERVER_PORT', 3000);

===============================================================================
SECURITY BEST PRACTICES IMPLEMENTED:
===============================================================================
1. ✓ Loads secrets from .env files (never hardcoded)
2. ✓ Validates required secrets are present
3. ✓ Never logs or prints secret values
4. ✓ Provides typed access to configuration values
5. ✓ Supports multiple environment files (.env, .env.local, etc.)
6. ✓ Validates configuration on startup
7. ✓ Type safety with JSDoc
8. ✓ Clear error messages when secrets are missing

===============================================================================
"""

const fs = require('fs');
const path = require('path');

/**
 * Custom error class for configuration errors
 */
class ConfigurationError extends Error {
    constructor(message) {
        super(message);
        this.name = 'ConfigurationError';
    }
}

/**
 * Secure configuration manager for loading environment variables from .env files.
 * 
 * Features:
 * - Loads from .env files with proper precedence
 * - Validates required secrets are present
 * - Provides type-safe access to configuration
 * - Never logs secret values
 * - Supports environment-specific configurations
 */
class SecureConfig {
    /**
     * Initialize configuration loader
     * @param {string[]} envFiles - List of .env file paths to load
     */
    constructor(envFiles = null) {
        this.envFiles = envFiles || this._getDefaultEnvFiles();
        this.loaded = false;
        this.config = {};
        
        this._loadEnvFiles();
        this._validateConfiguration();
    }

    /**
     * Get default .env file locations in order of precedence
     * Later files override earlier ones.
     * @returns {string[]} Array of .env file paths
     */
    _getDefaultEnvFiles() {
        const env = process.env.NODE_ENV || 'development';
        const baseDir = __dirname;

        return [
            path.join(baseDir, '.env'),
            path.join(baseDir, `.env.${env}`),
            path.join(baseDir, '.env.local'),
        ].filter(file => fs.existsSync(file));
    }

    /**
     * Load environment variables from .env files
     * @private
     */
    _loadEnvFiles() {
        if (this.envFiles.length === 0) {
            console.warn(
                '⚠️  No .env files found. Create .env file with required secrets. ' +
                'See .env.example for template.'
            );
            return;
        }

        this.envFiles.forEach(envFile => {
            try {
                const content = fs.readFileSync(envFile, 'utf8');
                this._parseEnvFile(content);
                console.log(`✓ Loaded configuration from: ${envFile}`);
            } catch (error) {
                throw new ConfigurationError(
                    `Failed to load ${envFile}: ${error.message}`
                );
            }
        });

        this.loaded = true;
    }

    /**
     * Parse .env file content into key-value pairs
     * @param {string} content - Content of .env file
     * @private
     */
    _parseEnvFile(content) {
        const lines = content.split('\n');

        lines.forEach(line => {
            // Skip empty lines and comments
            line = line.trim();
            if (!line || line.startsWith('#')) return;

            // Parse KEY=VALUE
            const [key, ...valueParts] = line.split('=');
            if (!key) return;

            let value = valueParts.join('=').trim();

            // Remove surrounding quotes if present
            if ((value.startsWith('"') && value.endsWith('"')) ||
                (value.startsWith("'") && value.endsWith("'"))) {
                value = value.slice(1, -1);
            }

            this.config[key.trim()] = value;
            process.env[key.trim()] = value;
        });
    }

    /**
     * Validate that all required secrets are present
     * @private
     */
    _validateConfiguration() {
        const env = process.env.NODE_ENV || 'development';
        const requiredSecrets = ['API_KEY', 'JWT_SECRET'];

        if (env === 'production') {
            // Production might require additional secrets
            const productionRequired = [
                'API_KEY',
                'JWT_SECRET',
                'DATABASE_URL',
            ];
            requiredSecrets.push(...productionRequired);
        }

        const missing = requiredSecrets.filter(
            secret => !process.env[secret] && !this.config[secret]
        );

        if (missing.length > 0) {
            throw new ConfigurationError(
                `Missing required environment variables in ${env} mode: ` +
                `${missing.join(', ')}\n` +
                `Instructions:\n` +
                `  1. Copy .env.example to .env\n` +
                `  2. Fill in actual values for: ${missing.join(', ')}\n` +
                `  3. Never commit .env to version control`
            );
        }
    }

    /**
     * Get a configuration value as string
     * @param {string} key - Environment variable name
     * @param {string} [defaultValue] - Default value if not found
     * @returns {string|null} Configuration value or default
     * @throws {ConfigurationError} If key is required and not found
     */
    get(key, defaultValue = null) {
        const value = process.env[key] || this.config[key] || defaultValue;
        return value || null;
    }

    /**
     * Get a secret value securely
     * @param {string} key - Environment variable name for the secret
     * @param {boolean} [required=false] - If true, raises error if secret is missing
     * @returns {string|null} Secret value
     * @throws {ConfigurationError} If required=true and secret is missing
     */
    getSecret(key, required = false) {
        const value = process.env[key] || this.config[key];

        if (!value && required) {
            throw new ConfigurationError(
                `Required secret '${key}' not found in environment. ` +
                `Please set it in your .env file.`
            );
        }

        return value || null;
    }

    /**
     * Get a configuration value as integer
     * @param {string} key - Environment variable name
     * @param {number} [defaultValue] - Default value if not found
     * @returns {number} Configuration value as integer
     * @throws {ConfigurationError} If value cannot be converted to integer
     */
    getInt(key, defaultValue = null) {
        const value = process.env[key] || this.config[key];

        if (!value) {
            if (defaultValue !== null) return defaultValue;
            throw new ConfigurationError(`Configuration '${key}' not found`);
        }

        const parsed = parseInt(value, 10);
        if (isNaN(parsed)) {
            throw new ConfigurationError(
                `Configuration '${key}' must be an integer, got: ${value}`
            );
        }

        return parsed;
    }

    /**
     * Get a configuration value as boolean
     * @param {string} key - Environment variable name
     * @param {boolean} [defaultValue=false] - Default value if not found
     * @returns {boolean} Configuration value as boolean
     */
    getBoolean(key, defaultValue = false) {
        const value = (process.env[key] || this.config[key] || String(defaultValue)).toLowerCase();
        return ['true', '1', 'yes', 'on'].includes(value);
    }

    /**
     * Get a configuration value as array (comma-separated)
     * @param {string} key - Environment variable name
     * @param {string} [delimiter=','] - Delimiter to split the value
     * @param {string[]} [defaultValue=[]] - Default array if not found
     * @returns {string[]} Configuration value split as array
     */
    getArray(key, delimiter = ',', defaultValue = []) {
        const value = process.env[key] || this.config[key];

        if (!value) {
            return defaultValue;
        }

        return value
            .split(delimiter)
            .map(item => item.trim())
            .filter(item => item.length > 0);
    }

    /**
     * Validate that required keys are present
     * @param {string[]} keys - List of required environment variable names
     * @throws {ConfigurationError} If any required key is missing
     */
    validateRequired(keys) {
        const missing = keys.filter(
            key => !(process.env[key] || this.config[key])
        );

        if (missing.length > 0) {
            throw new ConfigurationError(
                `Missing required configuration: ${missing.join(', ')}`
            );
        }
    }

    /**
     * Get all configuration (excluding secrets for safety)
     * @returns {object} Current configuration
     */
    getAll() {
        // Filter out sensitive keys
        const sensitiveKeys = [
            'API_KEY', 'SECRET_KEY', 'PASSWORD', 'TOKEN', 'JWT_SECRET',
            'DATABASE_PASSWORD', 'SMTP_PASSWORD', 'AWS_SECRET_ACCESS_KEY'
        ];

        const filtered = {};
        Object.entries(process.env).forEach(([key, value]) => {
            if (!sensitiveKeys.some(sensitive => key.includes(sensitive))) {
                filtered[key] = value;
            }
        });

        return filtered;
    }

    /**
     * String representation without exposing secrets
     * @returns {string} String representation
     */
    toString() {
        const env = process.env.NODE_ENV || 'development';
        return (
            `<SecureConfig ` +
            `environment=${env} ` +
            `loaded_files=${this.envFiles.length} ` +
            `validation_passed=true>`
        );
    }
}

/**
 * Initialize Express app with secure configuration
 * @param {object} app - Express application instance
 */
function initSecureConfig(app) {
    try {
        // Set Express configuration from environment
        app.set('secret', config.getSecret('JWT_SECRET', false));
        app.set('env', process.env.NODE_ENV || 'development');
        app.set('trust proxy', config.getBoolean('TRUST_PROXY', false));

        // Security headers middleware
        app.use((req, res, next) => {
            res.set('X-Content-Type-Options', 'nosniff');
            res.set('X-Frame-Options', 'DENY');
            res.set('X-XSS-Protection', '1; mode=block');
            next();
        });

        console.log('✓ Express app configured securely');
    } catch (error) {
        console.error(`Failed to initialize Express config: ${error.message}`);
        throw error;
    }
}

// ============================================================================
// SINGLETON INSTANCE
// ============================================================================
let config;

try {
    config = new SecureConfig();
    console.log('✓ Configuration loaded successfully');
} catch (error) {
    console.error(`✗ Configuration Error: ${error.message}`);
    process.exit(1);
}

// ============================================================================
// EXPORTS
// ============================================================================
module.exports = {
    config,
    ConfigurationError,
    SecureConfig,
    initSecureConfig,
};

// ============================================================================
// EXAMPLE USAGE
// ============================================================================
if (require.main === module) {
    try {
        console.log('Loading configuration...\n');

        // Access various configuration types
        const apiKey = config.getSecret('API_KEY', true);
        console.log('✓ API Key loaded (value hidden for security)');

        const port = config.getInt('PORT', 3000);
        console.log(`✓ Server Port: ${port}`);

        const debug = config.getBoolean('DEBUG', false);
        console.log(`✓ Debug Mode: ${debug}`);

        const paths = config.getArray('MONITORED_PATHS');
        console.log(`✓ Monitored Paths: ${paths.length > 0 ? paths.join(', ') : 'none'}`);

        console.log('\n✓ All configurations loaded successfully!');
    } catch (error) {
        console.error(`✗ Configuration Error: ${error.message}`);
        process.exit(1);
    }
}
